File: sdk/go/keys.go
package authsdk

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"sync"
	"time"

	"github.com/lestrrat-go/jwx/v2/jwk"
)

// keyCache holds the JWKS and refreshes it in the background.
//
// Reads never block on the network. That is the whole point: a request being
// verified must not wait on, or fail because of, an HTTP call to authsvc.
type keyCache struct {
	c *Client

	mu        sync.RWMutex
	set       jwk.Set
	fetchedAt time.Time
	lastErr   error

	// refreshing guards against piling up concurrent refreshes when the
	// service is slow or down.
	refreshing sync.Mutex
}

func newKeyCache(c *Client) *keyCache { return &keyCache{c: c} }

var ErrNoKeys = errors.New("authsdk: no verification keys available yet")

// get returns the cached key set. It never performs a fetch; if the set is
// stale it kicks off a background refresh and returns the stale set anyway,
// because a stale key still verifies a valid token correctly.
func (k *keyCache) get() (jwk.Set, error) {
	k.mu.RLock()
	set, at, err := k.set, k.fetchedAt, k.lastErr
	k.mu.RUnlock()

	if set == nil {
		// Cold and never successfully fetched. Trigger a refresh so the next
		// request can succeed, but do not wait for it.
		go k.refresh()
		if err != nil {
			return nil, fmt.Errorf("%w: %v", ErrNoKeys, err)
		}
		return nil, ErrNoKeys
	}

	age := time.Since(at)
	if age > k.c.cfg.JWKSRefresh {
		go k.refresh() // stale-while-revalidate
	}
	if max := k.c.cfg.JWKSMaxStale; max > 0 && age > max {
		return nil, fmt.Errorf("authsdk: key set is %s stale, past JWKSMaxStale", age.Round(time.Second))
	}
	return set, nil
}

// warm does one bounded, best-effort fetch at construction.
func (k *keyCache) warm() {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := k.fetch(ctx); err != nil {
		k.report(err)
	}
}

func (k *keyCache) run() {
	t := time.NewTicker(k.c.cfg.JWKSRefresh)
	defer t.Stop()
	for {
		select {
		case <-k.c.stop:
			return
		case <-t.C:
			k.refresh()
		}
	}
}

func (k *keyCache) refresh() {
	// Only one refresh at a time; the rest simply return.
	if !k.refreshing.TryLock() {
		return
	}
	defer k.refreshing.Unlock()

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	if err := k.fetch(ctx); err != nil {
		k.report(err)
	}
}

func (k *keyCache) fetch(ctx context.Context) error {
	body, err := k.c.getWithFallback(ctx, "/.well-known/jwks.json")
	if err != nil {
		k.mu.Lock()
		k.lastErr = err
		k.mu.Unlock()
		return err
	}

	set, err := jwk.Parse(body)
	if err != nil {
		k.mu.Lock()
		k.lastErr = err
		k.mu.Unlock()
		return fmt.Errorf("authsdk: parse jwks: %w", err)
	}
	if set.Len() == 0 {
		// Never replace a working key set with an empty one.
		err := errors.New("authsdk: jwks contained no keys")
		k.mu.Lock()
		k.lastErr = err
		k.mu.Unlock()
		return err
	}

	k.mu.Lock()
	k.set, k.fetchedAt, k.lastErr = set, time.Now(), nil
	k.mu.Unlock()
	return nil
}

func (k *keyCache) report(err error) {
	if k.c.cfg.OnError != nil {
		k.c.cfg.OnError(err)
	}
}

// getWithFallback tries BaseURL, then FallbackURL on a connection error or 5xx.
// A 4xx is a real answer and is not retried against the fallback.
func (c *Client) getWithFallback(ctx context.Context, path string) ([]byte, error) {
	body, err := c.getOnce(ctx, c.cfg.BaseURL+path)
	if err == nil {
		return body, nil
	}
	if c.cfg.FallbackURL == "" || !shouldFailOver(err) {
		return nil, err
	}
	body, ferr := c.getOnce(ctx, c.cfg.FallbackURL+path)
	if ferr != nil {
		return nil, fmt.Errorf("authsdk: primary failed (%v); fallback failed: %w", err, ferr)
	}
	return body, nil
}

// httpStatusError marks a non-2xx response so failover can distinguish "the
// server is broken" from "the server said no".
type httpStatusError struct {
	Status int
	Body   string
}

func (e *httpStatusError) Error() string {
	return fmt.Sprintf("authsdk: server returned %d: %s", e.Status, e.Body)
}

func shouldFailOver(err error) bool {
	var se *httpStatusError
	if errors.As(err, &se) {
		return se.Status >= 500
	}
	// Transport-level failure: connection refused, DNS, timeout.
	return true
}

func (c *Client) getOnce(ctx context.Context, url string) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	return c.do(req)
}

func errorIsNoKeys(err error) bool { return errors.Is(err, ErrNoKeys) }

File: sdk/rust/src/keys.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use jsonwebtoken::{Algorithm, DecodingKey};
use serde::Deserialize;
use tokio::sync::RwLock;

use crate::{Config, Error};

/// One key from the JWKS. authsvc publishes Ed25519 (OKP) keys only.
#[derive(Debug, Deserialize, Clone)]
pub(crate) struct Jwk {
    pub kid: String,
    /// Key type; authsvc publishes OKP. Kept for completeness even though the
    /// curve check below is what actually gates usability.
    #[serde(default)]
    #[allow(dead_code)]
    pub kty: String,
    #[serde(default)]
    pub crv: String,
    /// Base64url public key material.
    pub x: String,
}

#[derive(Debug, Deserialize)]
pub(crate) struct JwkSet {
    pub keys: Vec<Jwk>,
}

#[derive(Clone)]
pub(crate) struct CachedKey {
    pub kid: String,
    pub key: Arc<DecodingKey>,
}

struct Inner {
    keys: Vec<CachedKey>,
    fetched_at: Option<Instant>,
    last_err: Option<String>,
}

/// Holds the JWKS and refreshes it in the background.
///
/// Reads never block on the network. That is the whole point: verifying a
/// request must not wait on, or fail because of, an HTTP call to authsvc.
pub(crate) struct KeyCache {
    inner: RwLock<Inner>,
    refreshing: tokio::sync::Mutex<()>,
    cfg: Config,
    http: reqwest::Client,
}

impl KeyCache {
    pub(crate) fn new(cfg: Config, http: reqwest::Client) -> Self {
        Self {
            inner: RwLock::new(Inner { keys: Vec::new(), fetched_at: None, last_err: None }),
            refreshing: tokio::sync::Mutex::new(()),
            cfg,
            http,
        }
    }

    /// Returns the cached keys. Never fetches inline; a stale set is returned
    /// as-is because a stale key still verifies a valid token correctly.
    pub(crate) async fn get(self: &Arc<Self>) -> Result<Vec<CachedKey>, Error> {
        let (keys, fetched_at, last_err) = {
            let g = self.inner.read().await;
            (g.keys.clone(), g.fetched_at, g.last_err.clone())
        };

        let Some(at) = fetched_at else {
            // Cold and never successfully fetched. Kick off a refresh but do
            // not wait for it.
            self.spawn_refresh();
            return Err(Error::NoKeys(last_err.unwrap_or_default()));
        };

        let age = at.elapsed();
        if age > self.cfg.jwks_refresh {
            self.spawn_refresh(); // stale-while-revalidate
        }
        if let Some(max) = self.cfg.jwks_max_stale {
            if age > max {
                return Err(Error::KeysTooStale { age, limit: max });
            }
        }
        Ok(keys)
    }

    fn spawn_refresh(self: &Arc<Self>) {
        let me = Arc::clone(self);
        tokio::spawn(async move { me.refresh().await });
    }

    pub(crate) async fn refresh(self: &Arc<Self>) {
        // Only one refresh at a time; the rest simply return.
        let Ok(_guard) = self.refreshing.try_lock() else { return };
        if let Err(e) = self.fetch().await {
            let msg = e.to_string();
            self.inner.write().await.last_err = Some(msg.clone());
            if let Some(cb) = &self.cfg.on_error {
                cb(&msg);
            }
        }
    }

    pub(crate) async fn fetch(&self) -> Result<(), Error> {
        let body = self.get_with_fallback("/.well-known/jwks.json").await?;
        let set: JwkSet = serde_json::from_slice(&body).map_err(|e| Error::Jwks(e.to_string()))?;

        if set.keys.is_empty() {
            // Never replace a working key set with an empty one.
            return Err(Error::Jwks("jwks contained no keys".into()));
        }

        let mut out = Vec::with_capacity(set.keys.len());
        for k in set.keys {
            if !k.crv.is_empty() && k.crv != "Ed25519" {
                continue; // authsvc signs with EdDSA only
            }
            let key = DecodingKey::from_ed_components(&k.x)
                .map_err(|e| Error::Jwks(format!("key {}: {e}", k.kid)))?;
            out.push(CachedKey { kid: k.kid, key: Arc::new(key) });
        }
        if out.is_empty() {
            return Err(Error::Jwks("no usable Ed25519 keys in jwks".into()));
        }

        let mut g = self.inner.write().await;
        g.keys = out;
        g.fetched_at = Some(Instant::now());
        g.last_err = None;
        Ok(())
    }

    /// Tries base_url, then fallback_url on a connection error or 5xx.
    /// A 4xx is a real answer and is not retried against the fallback.
    async fn get_with_fallback(&self, path: &str) -> Result<Vec<u8>, Error> {
        let primary = self.get_once(&format!("{}{}", self.cfg.base_url, path)).await;
        match primary {
            Ok(b) => Ok(b),
            Err(e) => {
                let Some(fb) = &self.cfg.fallback_url else { return Err(e) };
                if !e.should_fail_over() {
                    return Err(e);
                }
                self.get_once(&format!("{fb}{path}")).await
            }
        }
    }

    async fn get_once(&self, url: &str) -> Result<Vec<u8>, Error> {
        let resp = self
            .http
            .get(url)
            .timeout(Duration::from_secs(10))
            .send()
            .await
            .map_err(|e| Error::Transport(e.to_string()))?;

        let status = resp.status();
        let body = resp.bytes().await.map_err(|e| Error::Transport(e.to_string()))?;
        if !status.is_success() {
            return Err(Error::Status { status: status.as_u16(), body: String::from_utf8_lossy(&body).into_owned() });
        }
        Ok(body.to_vec())
    }
}

/// Algorithm authsvc signs with.
pub(crate) const ALG: Algorithm = Algorithm::EdDSA;
